subprojects/language-groovy/src/main/java/org/gradle/api/tasks/compile/GroovyCompile.java

```diff
@@ -94,13 +133,21 @@ protected void compile() {
         return compiler;
     }
 
+    private FileCollection determineGroovyCompileClasspath() {
+        if (experimentalCompilationAvoidanceEnabled()) {
+            return compilerPluginClasspath.plus(getClasspath());
+        } else {
+            return getClasspath();
+        }
+    }
+
     private DefaultGroovyJavaJointCompileSpec createSpec() {
         DefaultGroovyJavaJointCompileSpec spec = new DefaultGroovyJavaJointCompileSpecFactory(compileOptions).create();
         spec.setSourceFiles(getSource());
         spec.setDestinationDir(getDestinationDir());
         spec.setWorkingDir(getProject().getProjectDir());
         spec.setTempDir(getTemporaryDir());
-        spec.setCompileClasspath(ImmutableList.copyOf(getClasspath()));
+        spec.setCompileClasspath(ImmutableList.copyOf(getClasspath().plus(determineGroovyCompileClasspath())));
```

Comment:
⭕️ Note that the API Groovy compiler actually has an `AST transformation classloader`, so we could actually use the smaller classpath for the ast transformation classloader instead of the full one. See https://github.com/gradle/gradle/blob/9d928bfdf9135154f7d484a6ac79e43b449e3fd3/subprojects/language-groovy/src/main/java/org/gradle/api/internal/tasks/compile/ApiGroovyCompiler.java#L110-L117

We should do this later, probably as part of #9516.

Reply:
:+1: